rdr_service/storage.py

```diff
@@ -370,3 +372,102 @@ def get_storage_provider():
     provider_class = StorageProvider.get_provider(default=default_provider)
     return provider_class()
 
+
+class GoogleCloudStorageCSVReader:
+    """
+    Read a CSV file from a bucket and yield rows or chunks of rows.
+    """
+    def __init__(self, cloud_csv_file, column_names:list=None, delimiter=','):
+        """
+        :param cloud_csv_file: bucket and file blob name of cloud csv file.
+        :param column_names: list of column names if first row does not have the col names.
+        """
+        if not cloud_csv_file or not isinstance(cloud_csv_file, str):
+            raise ValueError('Invalid cloud storage csv file name.')
+
+        self.cloud_csv_file = cloud_csv_file
+        self.column_names = column_names
+        self._delimiter = delimiter
+        self._tmp_file = None  # Temp file object handle.
+        self._strio = None  # StringIO object
+        self._header = None  # delimited string of column names.
+        self._reader = None  # csv.DictReader object.
+        self._batch_size = 1000  # number of rows from temp file to read.
+
+    @staticmethod
+    def _copy_blob_to_temp(cloud_csv_file):
+        """
+        Copy the cloud storage blob to the local temporary directory.
+        :param cloud_csv_file: bucket and file blob name of cloud csv file.
+        :return: temp file handle.
+        """
+        tmp_file = tempfile.NamedTemporaryFile(prefix='cloud_')
+        if GAE_PROJECT == 'localhost':
+            provider = LocalFilesystemStorageProvider()
+        else:
+            provider = GoogleCloudStorageProvider()
+        with provider.open(cloud_csv_file, 'rt') as csv_file:
+            chunk_size = 1000 * 1024 * 10  # 10MB chunk.
+            while True:
+                chunk = csv_file.read(chunk_size)
+                tmp_file.write(chunk.encode('utf-8') if isinstance(chunk, str) else chunk)
+                if not chunk or len(chunk) < chunk_size:
+                    break
+        tmp_file.seek(0)
+        return tmp_file
```

Comment:
suggestion: what do you think of using the Blob's `download_to_file` function and use Google's implementation for getting a file downloaded to tmp?

Reply:
I'm familiar with downloading the files this way, in the interest of time I want to go with something I know works correctly.